Objects/frameobject.c

```diff
@@ -835,66 +851,49 @@ frame_alloc(PyCodeObject *code, PyObject **localsarray)
         state->free_list = state->free_list->f_back;
         _Py_NewReference((PyObject *)f);
     }
-    f->f_localsptr = localsarray;
+    f->f_frame = frame;
     f->f_own_locals_memory = owns;
     return f;
 }
 
-int
-_PyFrame_TakeLocals(PyFrameObject *f)
+void
+_PyFrame_TakeLocals(PyFrameObject *f, InterpreterFrame *frame)
 {
     assert(f->f_own_locals_memory == 0);
-    assert(f->f_stackdepth == 0);
-    Py_ssize_t size = frame_nslots(f);
-    PyObject **copy = PyMem_Malloc(sizeof(PyObject *)*size);
-    if (copy == NULL) {
-        for (int i = 0; i < size; i++) {
-            PyObject *o = f->f_localsptr[i];
-            Py_XDECREF(o);
+    assert(frame->frame_obj == NULL);
+
+    f->f_own_locals_memory = 1;
+    f->f_frame = frame;
+    assert(f->f_back == NULL);
+    if (frame->previous != NULL) {
+        PyFrameObject *back = _PyFrame_GetFrameObject(frame->previous);
+        if (back == NULL) {
+            /* Memory error here. Nothing we can do about it */
+            PyErr_Clear();
+            _PyErr_WriteUnraisableMsg("Out of memory lazily allocating frame->f_back", NULL);
+        }
+        else {
+            f->f_back = (PyFrameObject *)Py_NewRef(back);
         }
-        PyErr_NoMemory();
-        return -1;
     }
-    for (int i = 0; i < size; i++) {
-        PyObject *o = f->f_localsptr[i];
-        copy[i] = o;
+    frame->previous = NULL;
```

Comment:
Hummmm, why this?

Reply:
The "back" chain changes from `previous` in the `InterpreterFrame` to `f_back` in the `PyFrameObject` at this point.
I've moved this into the `if (frame->previous != NULL)` block and added a comment.